src/stan_math_backend/Stan_math_code_gen.ml

```diff
@@ -657,12 +680,47 @@ let pp_register_map_rect_functors ppf p =
     (list ~sep:cut pp_register_functor)
     (List.mapi ~f:(fun i f -> (i + 1, f)) functors)
 
+let fun_used_in_reduce_sum p =
+  let rec find_functors_expr accum Expr.Fixed.({pattern; _}) =
+    match pattern with
+    | FunApp (StanLib, "reduce_sum", {pattern= Var f; _} :: _) -> f :: accum
+    | FunApp (StanLib, _, fk :: _) -> find_functors_expr accum fk
+    | _ -> accum
```

Comment:
This needs to be recursive for all patterns, not just the first arg of functions. Otherwise expressions like this will miss the second function:
```stan
target += reduce_sum(f, ...) + reduce_sum(g, ...);
```
I realize this all is copied from the above `map_rect`registration code. That code is equally broken; as far as I know `map_rect` doesn't even need the registration except when using MPI so I guess no one noticed.
The correct way to fold over an expression is, e.g. how `Transform_mir`'s `collect_vars_expr` does it. That function also illustrates the next point: `accum` should be a `String.Set.t` instead of a list. We don't care about the order and don't need duplicates.

Reply:
Are you sure here? I added a test case in test/integration/good/code-gen/reduce_sum_m3.stan 
`  real t1a = reduce_sum(f1, y1, 1) + reduce_sum(f1a, y1, 1);`

where f1a is not used anywhere else and it creates the reduce sum functor and the model compiles. 

I was aware of this issue  thats why the find functors for reduce sum is:
```ocaml
let rec find_functors_expr accum Expr.Fixed.({pattern; _}) =
    match pattern with
    | FunApp (StanLib, "reduce_sum", {pattern= Var f; _} :: _) -> f :: accum
    | FunApp (StanLib, _, fk :: _) -> find_functors_expr accum fk
    | _ -> accum
```

while map_rects is
```ocaml
let find_functors_expr accum Expr.Fixed.({pattern; _}) =
    match pattern with
    | FunApp (StanLib, "map_rect", {pattern= Var f; _} :: _) -> f :: accum
    | _ -> accum
  in
```
Maybe I am misunderstanding something?

Reply:
That doesn't make sense to me. The pattern for `reduce_sum(f1, ...) + reduce_sum(f1a,...)` is something like `FunApp (StanLib, "add", {...{Var "f1"}} :: {...{Var "f1a"}} :: [])`. Skipping the tail--as the recursive branch does--should hide "f1a". It didn't compile when I tried it.

`f1a_functor__` is [here](https://github.com/bstatcomp/stanc3/blob/5697210202b693f7a283a770eb36508952d03363/test/integration/good/code-gen/cpp.expected#L16753-L16761) but where is `f1a_rsfunctor__`?

Jenkins says all tests passed and I think that involves compiling all good test models so...I don't even know?

Reply:
Let me check again. This one does seem weird.

Reply:
Ideally, the fold would recurse to all expression in the program and you wouldn't have to do it manually. I'm not sure what the behavior of Program.fold is. I ended up defining my own program fold in the pedantic mode branch that does recurse over all expressions and statements.

If the fold doesn't recurse over all expressions, why are we only looking for reduce_sum inside of StanLib function arguments? Couldn't it appear anywhere?

Reply:
Figured this out. Going to try the set thing also.

Reply:
Bailed on the sets thing. Will to a refactor with sets together with the fix to the map_rect function after this PR is merged.